### PR TITLE
[90X] Allow multi-IOV input for Millepede alignment framework

### PR DESCRIPTION
**Pull request by ghellwig**
Removes shortcoming of the alignment framework which has not been able to update the alignables after an update of the tracker geometry at an IOV boundary.

Removes also temporary fix of #16135 and enables the usage of multi-IOV input for Millepede.

Added various protection against misconfiguration.
 - consistency check of the configuration during the setup of the alignment campaign
 - consistency checks in the C++ code to detect misconfiguration

The code has been used to derive the 2016 legacy alignment.

The general AlignmentProducer has the new features disabled by default. It is only enabled for Millepede. If other alignment algorithms want to use it a flag has to be set to `True`:
```
process.AlignmentProducer.enableAlignableUpdates = True
```

The code is based on top of #17408, i.e. to better digest the changes, it might be useful to first review and eventually approve #17408.

EDIT: A dedicated presentation on this update has been given at the [Tracker Alignment meeting on 01-Mar-2017](https://indico.cern.ch/event/607508/contributions/2493829/attachments/1420329/2176319/Alignment-MultiIOV_CMS-TkAl_20170301.pdf)

### CODE REVIEW

**Comment by ghellwig** [comment]
please test

**Comment by mmusich** [comment]
hi Gregor, @ghellwig 
```
----- Begin Fatal Exception 09-Feb-2017 17:40:40 CET-----------------------
An exception of category 'Configuration' occurred while
   [0] Constructing the EventProcessor
   [1] Constructing module: class=PCLTrackerAlProducer label='SiPixelAliMilleAlignmentProducer'
Exception Message:
MissingParameter: The required parameter 'enableAlignableUpdates' was not specified.
----- End Fatal Exception -------------------------------------------------
```
looks like the new parameter was somehow missed in the configuration here:

https://github.com/ghellwig/cmssw/blob/831c2e9d9d8a266b013b2d10c6ae899be0c381de/Alignment/CommonAlignmentProducer/python/ALCARECOPromptCalibProdSiPixelAli_cff.py#L76-L106

**Comment by ghellwig** [comment]
Hi Marco,
yes, I had to leave in a hurry yesterday and overlooked that I did not commit this change.
It's time to finalize the unification of the separate PCL and offline alignment producers...
Fortunately, this PR brings us also a bit closer to this goal, i.e. less duplication of code and configurations.

I'll push it later because I currently have problems at lxplus.

**Comment by ghellwig** [comment]
please test

**Comment by ghellwig** [comment]
@smuzaffar I just checked the clang results and found that a unit test within this PR does not pass there.
It happens because the script used for this unit test is taken from the release.
This PR fixes this script and consequently the test passes in the unit tests section, but it would be nice if the clang check could acknowledge possible local changes, too.

**Comment by smuzaffar** [comment]
@ghellwig , I noticed that new ```<test name=... command=.../>``` is also being run at the build time (which is a bug). I have a fix for that which I am testing right now. So please ignore the clang error in this PR. The unit test testPrepareInputDb actually ran fine when it was run after the build.

**Comment by ghellwig** [comment]
@smuzaffar I am fine with ignoring it, but just wanted to report it :)

**Comment by ghellwig** [comment]
please test

**Comment by ghellwig** [comment]
@smuzaffar @davidlange6 I have a naive question:
Shouldn't the list of added commits in this PR be updated after #17408 got merged? Or at least the diff should not show the changes introduced by #17408 anymore, right?

PS: This PR was based on #17408, and I thought github should be clever enough to detect that. The check-mark at https://github.com/cms-sw/cmssw/pull/17475/commits/3815b8d2d7044806b35f4ab0a1e046ff344c4b56 above also indicates that my assumption is not completely wrong.

**Comment by smuzaffar** [comment]
@ghellwig , ideally yes but no idea why github does not recognize this. I can check with the GH support team

**Comment by ghellwig** [comment]
@franzoni to ease the digestion of this PR:
A dedicated presentation on this update has been given at the [Tracker Alignment meeting on 01-Mar-2017](https://indico.cern.ch/event/607508/contributions/2493829/attachments/1420329/2176319/Alignment-MultiIOV_CMS-TkAl_20170301.pdf).

**Comment by franzoni** [comment]
pelase test
thanks @ghellwig

**Comment by ghellwig** [comment]
please test

File: Alignment/MillePedeAlignmentAlgorithm/python/mpslib/Mpslibclass.py [hunk 1]
```diff
@@ -99,18 +99,18 @@ def read_db(self, db_file_name = __default_db_file_name):
...
+            self.JOBHOST.append(parts[7].strip())
             self.JOBINCR.append(int(parts[8]))
-            self.JOBREMARK.append(parts[9])
-            self.JOBSP1.append(parts[10])
-            self.JOBSP2.append(parts[11])
-            self.JOBSP3.append(parts[12])
+            self.JOBREMARK.append(parts[9].strip())
+            self.JOBSP1.append(parts[10].strip())
+            self.JOBSP2.append(parts[11].strip())
+            self.JOBSP3.append(parts[12].strip())
```
Comment (franzoni) [inline Alignment/MillePedeAlignmentAlgorithm/python/mpslib/Mpslibclass.py]: hello @ghellwig 
how do you know which elements of parts[]
you're in need of removing blanks from ?

Comment (ghellwig) [inline Alignment/MillePedeAlignmentAlgorithm/python/mpslib/Mpslibclass.py]: Everything that is not converting to an integer remains a string and can have unintended whitespace at the left and right end.
The conversion to integers ignores such whitespace anyways.

**Comment by franzoni** [comment]
no changes in plots 2017 ttbar, 2016B data, runII2016 25 and 50 ns

**Comment by franzoni** [comment]
+1

**Comment by franzoni** [comment]
including this PR 900 important for alignment 
may we have this PR merged, please ?

**Comment by davidlange6** [comment]
+1

**Pull request merged**